### ISSUE TITLE
[suggestion] Ask for additional buffer context via user input

### ISSUE THREAD

**Issue by MrGrinst** [frankroeder/parrot.nvim#19]
There are cases where it would be nice for the LLM to have the context of another file I have open. I don't have a good idea yet on the implementation specifics but I'm wondering if you would be open to a PR that prompts the user for additional buffers to include as context for the LLM.

**Comment by frankroeder** [comment]
I am open to have something experimental on a different branch, guessing it would take a little bit more time to test this out and search for the best solution.

**Comment by eterps** [comment]
> you would be open to a PR that prompts the user for additional buffers to include as context for the LLM

You might want to check [https://github.com/paul-gauthier/aider](aider) for inspiration which does exactly that.
Unfortunately it's a separate TUI app, not a neovim plugin. Still a lot can be learned from it.

**Comment by MrGrinst** [comment]
@frankroeder @eterps what do you think about having a template variable like `{{userinput_buffers}}`. When that var is detected in the template, it would use fzf-lua to ask the user to select from open buffers.

**Comment by frankroeder** [comment]
@MrGrinst this is what I already have on my local branch. Currently,  just don't have the time to finish it.

**Comment by frankroeder** [comment]
@MrGrinst, how is it going with this feature? I guess the async stuff of fzf-lua is not so easy to handle.

**Comment by MrGrinst** [comment]
Yeah I haven't had time to finish it for that exact reason. The async stuff is beyond my Lua skills.

**Comment by eterps** [comment]
Last week I played around with this [nvim integration for aider](https://github.com/joshuavial/aider.nvim).

My first impression is that aider doesn't integrate very well in nvim, but I liked the feature of this plugin to automatically add all open buffers as a context.

It really makes sense as long as you have a set of files opened in buffers for the task at hand (otherwise you would need to delete some buffers first).

**Comment by frankroeder** [comment]
I have added the feature to include the content of all open buffers. Feel free to check out the branch for pull request #37.

Just to be clear, tools like Aider and Devin aim to "replace" the programmer's work. We focus on supporting the programmer through LLMs in the first place.

**Comment by eterps** [comment]
> I have added the feature to include the content of all open buffers.

Oooh nice, that looks incredibly useful. Definitely gonna check it out :+1: 

> We focus on supporting the programmer through LLMs

Yeah, I agree, no need to turn this into Aider/Devin. The way `parrot.nvim`'s templates work offer more fine grained control anyway.

**Comment by eterps** [comment]
@frankroeder , it works great. Very useful feature.

I noticed that I had to change the last part of the prompt to:

```
Please finish the code above carefully and logically.
Respond with the snippet of code that should be inserted, but absolutely no explanations or code block markers.
```

Otherwise Claude 3.5 would keep adding explanations or code blocks.

I'm wondering if this would be nice to have this for chats as well (instead of just for commands).

**Comment by MrGrinst** [comment]
This is great, I like this approach! Thanks @frankroeder. I'm thinking of adding a check to skip files larger than a certain size. Wanna avoid expensive calls (already had one go through that was 100k tokens 😮). Would you be open to that?

**Comment by MrGrinst** [comment]
Well it ended up being pretty hacky and probably making some assumptions based on my usage that might not be true for others. Not gonna open a PR but for reference here's the approach: https://github.com/frankroeder/parrot.nvim/commit/2b644d8da17c45626dff6d08a093865846b8a422

**Comment by frankroeder** [comment]
@MrGrinst, that is a very good suggestion. I am considering adding an optional confirmation for a specific approximate token limit. Thank you very much for your idea and implementation.

**Comment by sf8193** [comment]
checking in on this, would love to be able to give a list of files (or partial amount of files from buffers open) to be used. Often i have many buffers open but want help with only a couple of the files

**Comment by pluma9** [comment]
Check out [these works](https://github.com/Robitx/gp.nvim/issues/206#issuecomment-2343189529) in a fork of gp.nvim. I hope anyone who plans to implement this can get inspired from the existing works.

**Comment by frankroeder** [comment]
@pluma9 thank you for your suggestion. Your fork was indeed helpful. I am now working on something in #112

**Comment by frankroeder** [comment]
Solved with #112.
